protocols/autonatv2/src/server/handler/dial_back.rs

```diff
@@ -0,0 +1,148 @@
+    outbound: FuturesSet<ToBehaviour>,
+}
+
+impl Handler {
+    pub(crate) fn new() -> Self {
+        Self {
+            pending_nonce: VecDeque::new(),
+            requested_substream_nonce: VecDeque::new(),
+            outbound: FuturesSet::new(Duration::from_secs(10000), 2),
+        }
+    }
+}
+
+impl ConnectionHandler for Handler {
+    type FromBehaviour = FromBehaviour;
+    type ToBehaviour = ToBehaviour;
+    type InboundProtocol = DeniedUpgrade;
+    type OutboundProtocol = ReadyUpgrade<StreamProtocol>;
+    type InboundOpenInfo = ();
+    type OutboundOpenInfo = ();
+
+    fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
+        SubstreamProtocol::new(DeniedUpgrade, ())
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<
+        ConnectionHandlerEvent<Self::OutboundProtocol, Self::OutboundOpenInfo, Self::ToBehaviour>,
+    > {
+        if let Poll::Ready(result) = self.outbound.poll_unpin(cx) {
+            return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(
+                result
+                    .map_err(|timeout| io::Error::new(io::ErrorKind::TimedOut, timeout))
+                    .and_then(identity),
+            ));
+        }
+        if let Some(cmd) = self.pending_nonce.pop_front() {
+            self.requested_substream_nonce.push_back(cmd);
+            return Poll::Ready(ConnectionHandlerEvent::OutboundSubstreamRequest {
+                protocol: SubstreamProtocol::new(DIAL_BACK_UPGRADE, ()),
+            });
+        }
+        Poll::Pending
+    }
+
+    fn on_behaviour_event(&mut self, event: Self::FromBehaviour) {
+        self.pending_nonce.push_back(event);
+    }
+
+    fn on_connection_event(
+        &mut self,
+        event: ConnectionEvent<
+            Self::InboundProtocol,
+            Self::OutboundProtocol,
+            Self::InboundOpenInfo,
+            Self::OutboundOpenInfo,
+        >,
+    ) {
+        match event {
+            ConnectionEvent::FullyNegotiatedOutbound(FullyNegotiatedOutbound {
+                protocol, ..
+            }) => {
+                if let Some(cmd) = self.requested_substream_nonce.pop_front() {
+                    if self
+                        .outbound
+                        .try_push(perform_dial_back(protocol, cmd))
+                        .is_err()
+                    {
+                        tracing::warn!("Dial back dropped, too many requests in flight");
+                    }
+                } else {
+                    tracing::warn!("received dial back substream without nonce");
+                }
+            }
+            ConnectionEvent::DialUpgradeError(DialUpgradeError {
+                error: StreamUpgradeError::NegotiationFailed,
+                ..
+            })
+            | ConnectionEvent::DialUpgradeError(DialUpgradeError {
+                error: StreamUpgradeError::Timeout,
+                ..
+            }) => {
```

Comment:
```suggestion
            ConnectionEvent::DialUpgradeError(DialUpgradeError {
                error: StreamUpgradeError::NegotiationFailed | StreamUpgradeError::Timeout,
                ..
            }) => {
```

Reply:
Good suggestion, clippy should get that...